### PR TITLE
RavenDB-12911 Concurrency exception during spawning CandidateAmbassad…

### PR DESCRIPTION
**Pull request by karmeli87**
…ors can cause the cluster to hang

### CODE REVIEW

File: src/Raven.Server/Rachis/CandidateAmbassador.cs [hunk 1]
```diff
@@ -72,7 +72,7 @@ public void Dispose()
                     {
                         _engine.Log.Info(
                             $"CandidateAmbassador for {_tag}: Waited for a full second for thread {_candidateAmbassadorLongRunningWork.ManagedThreadId} " +
-                            $"({(_candidateAmbassadorLongRunningWork.Join(0) ? "running" : "finished")}) to finish, after the elections were {_candidate.ElectionResult}");
+                            $"({(_candidateAmbassadorLongRunningWork.Join(0) ? "finished" : "running")}) to finish, after the elections were {_candidate.ElectionResult}");
```
Comment (grisha-kotler) [inline src/Raven.Server/Rachis/CandidateAmbassador.cs:75]: finished to finish? running to finish?

File: src/Raven.Server/Rachis/Candidate.cs [hunk 2]
```diff
@@ -75,18 +75,7 @@ private void Run()
...
-                        }
-                        catch (ConcurrencyException)
-                        {
-                            foreach (var ambassador in _voters)
-                            {
-                                ambassador.Dispose();
-                            }
-                            return; // we lost the election, because someone else changed our state to follower
-                        }
+                        _engine.AppendStateDisposable(this, candidateAmbassador); // concurrency exception here will dispose the current candidate and it ambassadors
```
Comment (grisha-kotler) [inline src/Raven.Server/Rachis/Candidate.cs:78]: where is that disposed?

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Rachis/CandidateAmbassador.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Rachis/CandidateAmbassador.cs:75]: The issue was that we had an a polarity issue in the log statement.
If `Join` return `true`, we are finished, not running

File: src/Raven.Server/Rachis/Candidate.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/Rachis/Candidate.cs:78]: When the state of the engine changes, IIRC

**Review by ayende — APPROVED** [review]

File: src/Raven.Server/Rachis/CandidateAmbassador.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Server/Rachis/CandidateAmbassador.cs:75]: The message looks something like this:
`Waited for a full second for thread 123 (running) to finish`

**Review by grisha-kotler — APPROVED** [review]

**Pull request merged**